Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.Computability.Language
+
+/-!
+# Context-Free Grammar
+
+This file contains the definition of a Context-Free Grammar (CFG), which is a grammar that has a
+single nonterminal symbol on the left-hand side of each rule.
+Derivation by a grammar is inherently nondeterministic.
+-/
+
+/-- The type of symbols is the disjoint union of terminals `T` and nonterminals `N`. -/
+inductive Symbol (T : Type _) (N : Type _)
+  /-- Terminal symbols (of the same type as the language) -/
+  | terminal    (t : T) : Symbol T N
+  /-- Nonterminal symbols (must not be present at the end of word being generated) -/
+  | nonterminal (n : N) : Symbol T N
+  -- If we have `[BEq T]` and `[BEq N]` then `[BEq (Symbol T N)]` is added.
+  deriving BEq
```

Comment:
Is there a reason you've gone for `BEq` and not `DecidableEq`?

Reply:
I'll change it. Thanks!